# Import libraries and database connect

In [84]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sb
import os
import sqlite3

In [85]:
db_path = "nem_data.db"  # Change this to your actual database file path
conn = sqlite3.connect(db_path)

# Data view

In [2]:
df_unit_scada = pd.read_csv("PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202412010000.CSV", skiprows=1)

In [3]:
df_unit_scada.head()

I  DISPATCH  UNIT_SCADA    1       SETTLEMENTDATE     DUID  SCADAVALUE  \
0  D  DISPATCH  UNIT_SCADA  1.0  2024/12/01 00:05:00   ADPBA1        0.09   
1  D  DISPATCH  UNIT_SCADA  1.0  2024/12/01 00:05:00   ADPPV1        0.00   
2  D  DISPATCH  UNIT_SCADA  1.0  2024/12/01 00:05:00   AGLHAL        0.00   
3  D  DISPATCH  UNIT_SCADA  1.0  2024/12/01 00:05:00   AGLSOM        0.00   
4  D  DISPATCH  UNIT_SCADA  1.0  2024/12/01 00:05:00  ANGAST1        0.00   

           LASTCHANGED  
0  2024/12/01 00:00:11  
1  2024/12/01 00:00:11  
2  2024/12/01 00:00:11  
3  2024/12/01 00:00:11  
4  2024/12/01 00:00:11

In [4]:
df_unit_scada.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4048391 entries, 0 to 4048390
Data columns (total 8 columns):
 #   Column          Dtype  
---  ------          -----  
 0   I               object 
 1   DISPATCH        object 
 2   UNIT_SCADA      object 
 3   1               float64
 4   SETTLEMENTDATE  object 
 5   DUID            object 
 6   SCADAVALUE      float64
 7   LASTCHANGED     object 
dtypes: float64(2), object(6)
memory usage: 247.1+ MB


In [7]:
df_unit_scada.shape[0]

4048391

In [8]:
df_detail_summary = pd.read_csv("PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202412010000.CSV", skiprows=1)

In [9]:
df_detail_summary.head()

I  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY    7     DUID  \
0  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0   ADPBA1   
1  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   
2  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   
3  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   
4  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   

            START_DATE             END_DATE   DISPATCHTYPE CONNECTIONPOINTID  \
0  2024/09/06 00:00:00  2999/12/31 00:00:00  BIDIRECTIONAL            SMVE14   
1  2021/04/20 00:00:00  2021/06/29 00:00:00      GENERATOR            SMVE4D   
2  2021/06/29 00:00:00  2021/07/01 00:00:00      GENERATOR            SMVE4D   
3  2021/07/01 00:00:00  2021/11/09 00:00:00      GENERATOR            SMVE4D   
4  2021/11/09 00:00:00  2022/07/01 00:00:00      GENERATOR            SMVE4D   

  REGIONID  ... IS_AGGREGATED DISPATCHSUBTYPE ADG_ID  \
0      SA1  ...           1.0             NaN    NaN   
1      SA1  ...           1.0             NaN    NaN   
2      SA1  ...           1.0             NaN    NaN   
3      SA1  ...           1.0             NaN    NaN   
4      SA1  ...           1.0             NaN    NaN   

   LOAD_MINIMUM_ENERGY_PRICE LOAD_MAXIMUM_ENERGY_PRICE  LOAD_MIN_RAMP_RATE_UP  \
0                   -1013.53                  17736.74                    3.0   
1                        NaN                       NaN                    NaN   
2                        NaN                       NaN                    NaN   
3                        NaN                       NaN                    NaN   
4                        NaN                       NaN                    NaN   

   LOAD_MIN_RAMP_RATE_DOWN  LOAD_MAX_RAMP_RATE_UP LOAD_MAX_RAMP_RATE_DOWN  \
0                      3.0                    4.0                     4.0   
1                      NaN                    NaN                     NaN   
2                      NaN                    NaN                     NaN   
3                      NaN                    NaN                     NaN   
4                      NaN                    NaN                     NaN   

   SECONDARY_TLF  
0         1.0025  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 33 columns]

In [10]:
df_detail_summary.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13287 entries, 0 to 13286
Data columns (total 33 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   I                          13287 non-null  object 
 1   PARTICIPANT_REGISTRATION   13287 non-null  object 
 2   DUDETAILSUMMARY            13287 non-null  object 
 3   7                          13286 non-null  float64
 4   DUID                       13286 non-null  object 
 5   START_DATE                 13286 non-null  object 
 6   END_DATE                   13286 non-null  object 
 7   DISPATCHTYPE               13286 non-null  object 
 8   CONNECTIONPOINTID          13286 non-null  object 
 9   REGIONID                   13286 non-null  object 
 10  STATIONID                  13286 non-null  object 
 11  PARTICIPANTID              13286 non-null  object 
 12  LASTCHANGED                13286 non-null  object 
 13  TRANSMISSIONLOSSFACTOR     13286 non-null  flo

In [12]:
!pip install openpyxl

     |████████████████████████████████| 250 kB 4.2 MB/s eta 0:00:01


In [6]:
file_path = "NEM Registration and Exemption List.xlsx"
df_participants = pd.read_excel(file_path, sheet_name='PU and Scheduled Loads')

In [7]:
df_participants

Participant  \
0                   South Australian Water Corporation   
1                   South Australian Water Corporation   
2                   South Australian Water Corporation   
3                   South Australian Water Corporation   
4                   South Australian Water Corporation   
..                                                 ...   
528  Tailem Bend II Project Company Pty Ltd as trus...   
529                          AGL Macquarie Pty Limited   
530                      AGL SA Generation Pty Limited   
531  Capital Battery Pty Ltd as Trustee for Capital...   
532                    Hornsdale Power Reserve Pty Ltd   

                                     Station Name Region       Dispatch Type  \
0                     Adelaide Desalination Plant    SA1     Generating Unit   
1                     Adelaide Desalination Plant    SA1                Load   
2                     Adelaide Desalination Plant    SA1     Generating Unit   
3                     Adelaide Desalination Plant    SA1     Generating Unit   
4                     Adelaide Desalination Plant    SA1     Generating Unit   
..                                            ...    ...                 ...   
528  Tailem Bend 2 Hybrid Renewable Power Station    SA1  Bidirectional Unit   
529     Broken Hill Battery Energy Storage System   NSW1  Bidirectional Unit   
530                           Torrens Island BESS    SA1  Bidirectional Unit   
531                               Capital Battery   NSW1                Load   
532                       Hornsdale Power Reserve    SA1  Bidirectional Unit   

    Category  Classification Fuel Source - Primary Fuel Source - Descriptor  \
0     Market       Scheduled       Battery storage                     Grid   
1     Market       Scheduled       Battery storage                     Grid   
2     Market   Non-Scheduled                 Hydro                    Water   
3     Market  Semi-Scheduled                 Solar                    Solar   
4     Market   Non-Scheduled                 Solar                    Solar   
..       ...             ...                   ...                      ...   
528   Market       Scheduled       Battery storage                     Grid   
529   Market       Scheduled       Battery storage                     Grid   
530   Market       Scheduled       Battery storage                     Grid   
531   Market       Scheduled       Battery storage                     Grid   
532   Market       Scheduled       Battery storage                     Wind   

    Technology Type - Primary      Technology Type - Descriptor     Units  \
0                     Storage              Battery and Inverter       1-3   
1                     Storage              Battery and Inverter       1-3   
2                   Renewable                      Run of River       1-2   
3                   Renewable  Photovoltaic Tracking Flat panel  1-4, 5-9   
4                   Renewable           Photovoltaic Flat panel       1-2   
..                        ...                               ...       ...   
528                   Storage              Battery and Inverter      1-14   
529                   Storage              Battery and Inverter      1-44   
530                   Storage              Battery and Inverter     1-109   
531                   Storage              Battery and Inverter      1-33   
532                   Storage              Battery and Inverter     1-294   

    Aggregation     DUID Reg Cap generation (MW) Max Cap generation (MW)  \
0             Y  ADPBA1G                    7.76                    6.15   
1             Y  ADPBA1L                    7.76                    6.15   
2             Y   ADPMH1                    1.44                       1   
3             Y   ADPPV1                   24.75                      19   
4             Y   ADPPV2                     0.2                     0.2   
..          ...      ...                     ...       

In [19]:
df_participants['Fuel Source - Descriptor'].unique()

array(['Grid', 'Water', 'Solar', 'Diesel', 'Waste Coal Mine Gas', 'Wind',
       'Natural Gas', nan, 'Black Coal', 'Sewerage / Waste Water',
       'Coal Seam Methane', 'Bagasse', 'Landfill Methane / Landfill Gas',
       'Biogas - sludge', 'Solar ', 'Natural Gas / Diesel', 'Ethane',
       'Brown Coal', 'solar', 'Kerosene', '-', 'Natural Gas / Fuel Oil',
       'Bagasse and Diesel'], dtype=object)

In [5]:
# Define start and end dates
start_date = pd.to_datetime("2024-08-01")
end_date = pd.to_datetime("2025-01-01")

# Create an empty list to store DataFrames
df_list = []

# Loop through each month from Aug 2024 to Jan 2025
current_date = start_date
while current_date <= end_date:
    year_month = current_date.strftime("%Y%m")  # Format YYYYMM

    # Construct the file path
    file_path = f"extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#{year_month}010000.CSV"

    # Check if file exists before reading
    if os.path.exists(file_path):
        df_temp = pd.read_csv(file_path, skiprows=1)  # Skip the first row
        df_list.append(df_temp)
        print(f"✅ Loaded data from {file_path}, rows: {len(df_temp)}")
    else:
        print(f"⚠️ File not found: {file_path}")

    # Move to the next month
    current_date += pd.DateOffset(months=1)

# Concatenate all DataFrames
if df_list:
    df_demand = pd.concat(df_list, ignore_index=True)
    print(f"🎉 Successfully concatenated {len(df_demand)} rows from all files!")
else:
    df_demand = pd.DataFrame()  # Empty DataFrame if no files were found
    print("❌ No data files found for the given period.")

# Display the first few rows
df_demand.head()

✅ Loaded data from extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#202408010000.CSV, rows: 7441
✅ Loaded data from extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#202409010000.CSV, rows: 7201
✅ Loaded data from extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#202410010000.CSV, rows: 7441
✅ Loaded data from extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#202411010000.CSV, rows: 7201
✅ Loaded data from extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#202412010000.CSV, rows: 7441
✅ Loaded data from extracted_csv_demand/PUBLIC_ARCHIVE#DEMANDOPERATIONALACTUAL#FILE01#202501010000.CSV, rows: 7441
🎉 Successfully concatenated 44166 rows from all files!


I  OPERATIONAL_DEMAND  ACTUAL    3    INTERVAL_DATETIME REGIONID  \
0  D  OPERATIONAL_DEMAND  ACTUAL  3.0  2024/08/01 00:30:00     NSW1   
1  D  OPERATIONAL_DEMAND  ACTUAL  3.0  2024/08/01 00:30:00     QLD1   
2  D  OPERATIONAL_DEMAND  ACTUAL  3.0  2024/08/01 00:30:00      SA1   
3  D  OPERATIONAL_DEMAND  ACTUAL  3.0  2024/08/01 00:30:00     TAS1   
4  D  OPERATIONAL_DEMAND  ACTUAL  3.0  2024/08/01 00:30:00     VIC1   

   OPERATIONAL_DEMAND.1          LASTCHANGED  OPERATIONAL_DEMAND_ADJUSTMENT  \
0                8813.0  2024/08/01 00:30:02                            0.0   
1                6148.0  2024/08/01 00:30:02                            0.0   
2                1603.0  2024/08/01 00:30:02                            0.0   
3                1322.0  2024/08/01 00:30:02                            0.0   
4                5979.0  2024/08/01 00:30:02                            0.0   

   WDR_ESTIMATE  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0

In [76]:
# Define start and end dates
start_date = pd.to_datetime("2024-08-01")
end_date = pd.to_datetime("2025-01-31")

# Create an empty list to store DataFrames
df_list = []

# Loop through each month from Aug 2024 to Jan 2025
current_date = start_date
while current_date <= end_date:
    year_month = current_date.strftime("%Y%m")  # Format YYYYMM

    # Construct the file path
    file_path = f"extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#{year_month}010000.CSV"

    # Check if file exists before reading
    if os.path.exists(file_path):
        df_temp = pd.read_csv(file_path, skiprows=1)  # Skip the first row
        df_list.append(df_temp)
        print(f"✅ Loaded data from {file_path}, rows: {len(df_temp)}")
    else:
        print(f"⚠️ File not found: {file_path}")

    # Move to the next month
    current_date += pd.DateOffset(months=1)

# Concatenate all DataFrames
if df_list:
    df_unit_scada = pd.concat(df_list, ignore_index=True)
    print(f"🎉 Successfully concatenated {len(df_unit_scada)} rows from all files!")
else:
    df_unit_scada = pd.DataFrame()  # Empty DataFrame if no files were found
    print("❌ No data files found for the given period.")

# Display the first few rows
df_unit_scada.head()

✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202408010000.CSV, rows: 4136178
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202409010000.CSV, rows: 4034259
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202410010000.CSV, rows: 4082274


/var/folders/b6/bmslj24d11n41826_8m7yb400000gn/T/ipykernel_1092/4136954492.py:18: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  df_temp = pd.read_csv(file_path, skiprows=1)  # Skip the first row


✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202411010000.CSV, rows: 3902492
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202412010000.CSV, rows: 4048391
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DISPATCH_UNIT_SCADA#FILE01#202501010000.CSV, rows: 4058257
🎉 Successfully concatenated 24261851 rows from all files!


I  DISPATCH  UNIT_SCADA       1       SETTLEMENTDATE     DUID  SCADAVALUE  \
0  D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00  ADPBA1G     0.00000   
1  D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00  ADPBA1L     2.90700   
2  D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00   ADPPV1     0.00000   
3  D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00   AGLHAL     0.00000   
4  D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00   AGLSOM    19.82500   

  LASTCHANGED  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [48]:
df_unit_scada.tail()

I       DISPATCH  UNIT_SCADA    1       SETTLEMENTDATE   DUID  \
24261846  D       DISPATCH  UNIT_SCADA  1.0  2025/02/01 00:00:00  YWPS1   
24261847  D       DISPATCH  UNIT_SCADA  1.0  2025/02/01 00:00:00  YWPS2   
24261848  D       DISPATCH  UNIT_SCADA  1.0  2025/02/01 00:00:00  YWPS3   
24261849  D       DISPATCH  UNIT_SCADA  1.0  2025/02/01 00:00:00  YWPS4   
24261850  C  END OF REPORT     4058259  NaN                  NaN    NaN   

          SCADAVALUE          LASTCHANGED  
24261846    280.2500  2025/01/31 23:55:11  
24261847    279.7500  2025/01/31 23:55:11  
24261848    273.9375  2025/01/31 23:55:11  
24261849      0.0000  2025/01/31 23:55:11  
24261850         NaN                  NaN

In [49]:
# Define start and end dates
start_date = pd.to_datetime("2024-08-01")
end_date = pd.to_datetime("2025-01-31")

# Create an empty list to store DataFrames
df_list = []

# Loop through each month from Aug 2024 to Jan 2025
current_date = start_date
while current_date <= end_date:
    year_month = current_date.strftime("%Y%m")  # Format YYYYMM

    # Construct the file path
    file_path = f"extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#{year_month}010000.CSV"

    # Check if file exists before reading
    if os.path.exists(file_path):
        df_temp = pd.read_csv(file_path, skiprows=1)  # Skip the first row
        df_list.append(df_temp)
        print(f"✅ Loaded data from {file_path}, rows: {len(df_temp)}")
    else:
        print(f"⚠️ File not found: {file_path}")

    # Move to the next month
    current_date += pd.DateOffset(months=1)

# Concatenate all DataFrames
if df_list:
    df_summary = pd.concat(df_list, ignore_index=True)
    print(f"🎉 Successfully concatenated {len(df_summary)} rows from all files!")
else:
    df_summary = pd.DataFrame()  # Empty DataFrame if no files were found
    print("❌ No data files found for the given period.")

# Display the first few rows
df_summary.head()

✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202408010000.CSV, rows: 13246
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202409010000.CSV, rows: 13260
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202410010000.CSV, rows: 13269
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202411010000.CSV, rows: 13276
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202412010000.CSV, rows: 13287
✅ Loaded data from extracted_csv/PUBLIC_ARCHIVE#DUDETAILSUMMARY#FILE01#202501010000.CSV, rows: 13295
🎉 Successfully concatenated 79633 rows from all files!


I  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY    7     DUID  \
0  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0   ADPBA1   
1  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   
2  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   
3  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   
4  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ADPBA1G   

            START_DATE             END_DATE   DISPATCHTYPE CONNECTIONPOINTID  \
0  2024/09/06 00:00:00  2999/12/31 00:00:00  BIDIRECTIONAL            SMVE14   
1  2021/04/20 00:00:00  2021/06/29 00:00:00      GENERATOR            SMVE4D   
2  2021/06/29 00:00:00  2021/07/01 00:00:00      GENERATOR            SMVE4D   
3  2021/07/01 00:00:00  2021/11/09 00:00:00      GENERATOR            SMVE4D   
4  2021/11/09 00:00:00  2022/07/01 00:00:00      GENERATOR            SMVE4D   

  REGIONID  ... IS_AGGREGATED DISPATCHSUBTYPE ADG_ID  \
0      SA1  ...           1.0             NaN    NaN   
1      SA1  ...           1.0             NaN    NaN   
2      SA1  ...           1.0             NaN    NaN   
3      SA1  ...           1.0             NaN    NaN   
4      SA1  ...           1.0             NaN    NaN   

   LOAD_MINIMUM_ENERGY_PRICE LOAD_MAXIMUM_ENERGY_PRICE  LOAD_MIN_RAMP_RATE_UP  \
0                   -1013.53                  17736.74                    3.0   
1                        NaN                       NaN                    NaN   
2                        NaN                       NaN                    NaN   
3                        NaN                       NaN                    NaN   
4                        NaN                       NaN                    NaN   

   LOAD_MIN_RAMP_RATE_DOWN  LOAD_MAX_RAMP_RATE_UP LOAD_MAX_RAMP_RATE_DOWN  \
0                      3.0                    4.0                     4.0   
1                      NaN                    NaN                     NaN   
2                      NaN                    NaN                     NaN   
3                      NaN                    NaN                     NaN   
4                      NaN                    NaN                     NaN   

   SECONDARY_TLF  
0         1.0025  
1            NaN  
2            NaN  
3            NaN  
4            NaN  

[5 rows x 33 columns]

# Query data 

In [19]:
sql_scada = """SELECT * FROM DISPATCH_UNIT_SCADA
WHERE SETTLEMENTDATE LIKE '2024/08%' OR SETTLEMENTDATE LIKE '2024/09%' 
OR SETTLEMENTDATE LIKE '2024/10%' OR SETTLEMENTDATE LIKE '2024/11%'
OR SETTLEMENTDATE LIKE '2024/12%' OR SETTLEMENTDATE LIKE '2025/01%';
"""

sql_summary = """
SELECT * FROM DUDETAILSUMMARY
WHERE START_DATE LIKE '2024/08%' OR START_DATE LIKE '2024/09%' 
   OR START_DATE LIKE '2024/10%' OR START_DATE LIKE '2024/11%' 
   OR START_DATE LIKE '2024/12%' OR START_DATE LIKE '2025/01%';
"""

In [21]:
# Read data into Pandas DataFrames
df_scada = pd.read_sql_query(sql_scada, conn)
df_summary = pd.read_sql_query(sql_summary, conn)

# Close database connection
conn.close()

# Display the number of rows retrieved
print(f"✅ Loaded {len(df_scada)} rows from DISPATCH_UNIT_SCADA")
print(f"✅ Loaded {len(df_summary)} rows from DUDETAILSUMMARY")

✅ Loaded 20204066 rows from DISPATCH_UNIT_SCADA
✅ Loaded 62 rows from DUDETAILSUMMARY


In [22]:
df_scada.head()

I  DISPATCH  UNIT_SCADA    1       SETTLEMENTDATE      DUID  SCADAVALUE  \
0  D  DISPATCH  UNIT_SCADA  1.0  2024/08/01 00:00:00  CAPTL_WF   23.394888   
1  D  DISPATCH  UNIT_SCADA  1.0  2024/08/01 00:00:00   BARCSF1    0.100000   
2  D  DISPATCH  UNIT_SCADA  1.0  2024/08/01 00:00:00  BUTLERSG    7.699998   
3  D  DISPATCH  UNIT_SCADA  1.0  2024/08/01 00:00:00     BWTR1    4.172680   
4  D  DISPATCH  UNIT_SCADA  1.0  2024/08/01 00:00:00  CHALLHWF    0.000000   

  LASTCHANGED  
0        None  
1        None  
2        None  
3        None  
4        None

In [23]:
df_summary.head()

I  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY    7      DUID  \
0  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0    ADPBA1   
1  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0  ASDLBY01   
2  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0   BLYTHB1   
3  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0   BOWWBA1   
4  D  PARTICIPANT_REGISTRATION  DUDETAILSUMMARY  7.0   CBWWBA1   

            START_DATE             END_DATE   DISPATCHTYPE CONNECTIONPOINTID  \
0  2024/09/06 00:00:00  2999/12/31 00:00:00  BIDIRECTIONAL            SMVE14   
1  2024/08/13 00:00:00  2999/12/31 00:00:00           LOAD          QASDAL01   
2  2024/08/20 00:00:00  2999/12/31 00:00:00  BIDIRECTIONAL            SBYW1B   
3  2024/09/06 00:00:00  2999/12/31 00:00:00  BIDIRECTIONAL            SPGW6B   
4  2024/09/06 00:00:00  2999/12/31 00:00:00  BIDIRECTIONAL            SMVE15   

  REGIONID  ... IS_AGGREGATED DISPATCHSUBTYPE ADG_ID  \
0      SA1  ...           1.0            None   None   
1     QLD1  ...           NaN            None   None   
2      SA1  ...           1.0            None   None   
3      SA1  ...           1.0            None   None   
4      SA1  ...           1.0            None   None   

   LOAD_MINIMUM_ENERGY_PRICE LOAD_MAXIMUM_ENERGY_PRICE  LOAD_MIN_RAMP_RATE_UP  \
0                   -1013.53                  17736.74                    3.0   
1                        NaN                       NaN                    NaN   
2                    -986.60                  17265.50                   64.0   
3                    -991.40                  17349.44                    2.0   
4                   -1047.02                  18322.70                    2.0   

   LOAD_MIN_RAMP_RATE_DOWN  LOAD_MAX_RAMP_RATE_UP LOAD_MAX_RAMP_RATE_DOWN  \
0                      3.0                    4.0                     4.0   
1                      NaN                    NaN                     NaN   
2                     64.0                   80.0                    80.0   
3                      2.0                    2.0                     2.0   
4                      2.0                    2.0                     2.0   

   SECONDARY_TLF  
0         1.0025  
1            NaN  
2         0.9711  
3         1.0004  
4         1.0025  

[5 rows x 33 columns]

In [25]:
df_scada.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20204066 entries, 0 to 20204065
Data columns (total 8 columns):
 #   Column          Dtype  
---  ------          -----  
 0   I               object 
 1   DISPATCH        object 
 2   UNIT_SCADA      object 
 3   1               float64
 4   SETTLEMENTDATE  object 
 5   DUID            object 
 6   SCADAVALUE      float64
 7   LASTCHANGED     object 
dtypes: float64(2), object(6)
memory usage: 1.2+ GB


# EDA 

## 1D Exploration

<div style="background-color: lightblue; padding: 10px;">

The data used for this EDA is now from 4 sources:
-  DISPATCH_UNIT_SCADA for power generation.
-  DUDETAILSUMMARY for dispatchable unit details.
-  DEMANDOPERATIONALACTUAL for demand data.
-  Data from "NEM Registration and Exemption List.xlsx" file for energy types.

Data used is from Aug 2024 to jan 2025.

</div>

### power generation and registered participants

#### SCADAVALUE

In [50]:
df_unit_scada["SETTLEMENTDATE"] = pd.to_datetime(df_unit_scada["SETTLEMENTDATE"], format="%Y/%m/%d %H:%M:%S")

In [51]:
df_unit_scada.SETTLEMENTDATE.min()

Timestamp('2024-08-01 00:05:00')

In [52]:
df_unit_scada.SETTLEMENTDATE.max()

Timestamp('2025-02-01 00:00:00')

In [53]:
df_unit_scada.I.unique()

array(['D', 'C'], dtype=object)

In [55]:
df_unit_scada = df_unit_scada[df_unit_scada['I'] == "D"]

In [56]:
df_unit_scada.count()

I                 24261845
DISPATCH          24261845
UNIT_SCADA        24261845
1                 24261845
SETTLEMENTDATE    24261845
DUID              24261845
SCADAVALUE        24261845
LASTCHANGED       11172459
dtype: int64

In [62]:
pd.options.display.float_format = "{:.5f}".format
df_unit_scada.describe()

1                 SETTLEMENTDATE     SCADAVALUE
count 24261845.00000                       24261845 24261845.00000
mean         1.00000  2024-10-31 12:12:11.041279744       45.83084
min          1.00000            2024-08-01 00:05:00     -257.00000
25%          1.00000            2024-09-15 10:55:00        0.00000
50%          1.00000            2024-10-31 01:40:00        0.00000
75%          1.00000            2024-12-16 16:20:00       30.60000
max          1.00000            2025-02-01 00:00:00     1783.01001
std          0.00000                            NaN      111.89076

In [61]:
df_unit_scada[df_unit_scada['SCADAVALUE'] < 0]

I  DISPATCH  UNIT_SCADA    1      SETTLEMENTDATE      DUID  \
22        D  DISPATCH  UNIT_SCADA 1.00 2024-08-01 00:05:00  BBATTERY   
27        D  DISPATCH  UNIT_SCADA 1.00 2024-08-01 00:05:00     BDL02   
29        D  DISPATCH  UNIT_SCADA 1.00 2024-08-01 00:05:00      BHB1   
74        D  DISPATCH  UNIT_SCADA 1.00 2024-08-01 00:05:00  CHBESSG1   
81        D  DISPATCH  UNIT_SCADA 1.00 2024-08-01 00:05:00    CLOVER   
...      ..       ...         ...  ...                 ...       ...   
24261779  D  DISPATCH  UNIT_SCADA 1.00 2025-02-01 00:00:00     TNPS1   
24261786  D  DISPATCH  UNIT_SCADA 1.00 2025-02-01 00:00:00  TULLYSM1   
24261796  D  DISPATCH  UNIT_SCADA 1.00 2025-02-01 00:00:00      VBB1   
24261808  D  DISPATCH  UNIT_SCADA 1.00 2025-02-01 00:00:00   WALGRV1   
24261835  D  DISPATCH  UNIT_SCADA 1.00 2025-02-01 00:00:00    WTAHB1   

          SCADAVALUE          LASTCHANGED  
22             -0.47                  NaN  
27             -0.24                  NaN  
29             -1.44                  NaN  
74             -0.21                  NaN  
81             -0.01                  NaN  
...              ...                  ...  
24261779       -1.06  2025/01/31 23:55:11  
24261786       -0.30  2025/01/31 23:55:11  
24261796       -3.80  2025/01/31 23:55:11  
24261808       -1.74  2025/01/31 23:55:11  
24261835       -7.95  2025/01/31 23:55:11  

[898904 rows x 8 columns]

In [63]:
df_unit_scada[df_unit_scada['SCADAVALUE'] == 0]

I  DISPATCH  UNIT_SCADA       1      SETTLEMENTDATE      DUID  \
0         D  DISPATCH  UNIT_SCADA 1.00000 2024-08-01 00:05:00   ADPBA1G   
2         D  DISPATCH  UNIT_SCADA 1.00000 2024-08-01 00:05:00    ADPPV1   
3         D  DISPATCH  UNIT_SCADA 1.00000 2024-08-01 00:05:00    AGLHAL   
5         D  DISPATCH  UNIT_SCADA 1.00000 2024-08-01 00:05:00   ANGAST1   
6         D  DISPATCH  UNIT_SCADA 1.00000 2024-08-01 00:05:00     ARWF1   
...      ..       ...         ...     ...                 ...       ...   
24261838  D  DISPATCH  UNIT_SCADA 1.00000 2025-02-01 00:00:00    YABULU   
24261839  D  DISPATCH  UNIT_SCADA 1.00000 2025-02-01 00:00:00   YABULU2   
24261841  D  DISPATCH  UNIT_SCADA 1.00000 2025-02-01 00:00:00  YARANSF1   
24261843  D  DISPATCH  UNIT_SCADA 1.00000 2025-02-01 00:00:00    YATSF1   
24261849  D  DISPATCH  UNIT_SCADA 1.00000 2025-02-01 00:00:00     YWPS4   

          SCADAVALUE          LASTCHANGED  
0            0.00000                  NaN  
2            0.00000                  NaN  
3            0.00000                  NaN  
5            0.00000                  NaN  
6            0.00000                  NaN  
...              ...                  ...  
24261838     0.00000  2025/01/31 23:55:11  
24261839     0.00000  2025/01/31 23:55:11  
24261841     0.00000  2025/01/31 23:55:11  
24261843     0.00000  2025/01/31 23:55:11  
24261849     0.00000  2025/01/31 23:55:11  

[11582746 rows x 8 columns]

In [77]:
df_unit_scada = df_unit_scada[df_unit_scada["SCADAVALUE"] >= 0]

<div style="background-color: lightblue; padding: 10px;">

There are around 0.89M rows of negative values for power generation 0.89M/24.2M rows. My solution for this is dropping these values, because they may be errors when collecting data from generators.

There are around 11.6M/24.2 rows (nearly 50% of data are 0) have power generation of 0MW. The reasons behind may be that the units did not collect data at that time point or they rounded the data to 0 because data is too small.

Maximum values for power generation of an unit is 1783 MW.
</div>

#### number of units

In [78]:
duid_lst = df_unit_scada.DUID.unique()

In [79]:
len(duid_lst)

500

In [80]:
registered_duid_lst = df_participants.DUID.unique()

In [81]:
len(registered_duid_lst)

525

<div style="background-color: lightblue; padding: 10px;">

There are 500 units recorded data in UNIT_SCADA_VALUE.

There are 525 units registered in the participants file.

Now, let's move to explore is there any units that not in the registered participant file.
</div>

In [82]:
diff_lst = list(set(duid_lst) - set(registered_duid_lst))
print(diff_lst)  

['HPRL1', 'GANNBG1', 'BBATTERY', 'RT_SA3', 'PIONEER', 'WALGRVG1', 'CBWWBA1L', 'DG_VIC1', 'GANNBL1', 'RT_NSW3', 'RT_SA4', 'RIVNBG2', 'RT_SA1', 'RT_VIC12', 'RT_VIC8', 'RT_VIC9', 'CAPBES1G', 'RT_VIC2', 'PORTWF', 'CBWWBA1G', 'BBATRYL1', 'RT_SA5', 'DG_QLD1', 'CHBESSG1', 'CAPBES1L', 'RT_VIC5', 'RT_VIC11', 'RT_SA6', 'LBBL1', 'RT_NSW1', 'DG_TAS1', 'BULBESG1', 'CHBESSL1', 'WDBESSL1', 'RT_NSW2', 'RT_VIC3', 'RT_VIC7', 'TB2BL1', 'TB2BG1', 'DG_SA1', 'RESS1G', 'HVWWBA1L', 'RT_VIC1', 'RT_NSW4', 'PIBESSG1', 'BOWWBA1G', 'RT_TAS1', 'RT_QLD1', 'WALGRVL1', 'DPNTBL1', 'RT_VIC10', 'VBBL1', 'BALBL1', 'RESS1L', 'RT_NSW6', 'RT_NSW5', 'BULBESL1', 'DPNTBG1', 'DG_NSW1', 'PIBESSL1', 'RT_VIC6', 'WDBESSG1', 'BALBG1', 'BOWWBA1L', 'HPRG1', 'LBBG1', 'RT_SA2', 'HVWWBA1G', 'RIVNBL2', 'RT_VIC4', 'VBBG1']


In [83]:
df_unit_scada[df_unit_scada['DUID'].isin(diff_lst)]

I  DISPATCH  UNIT_SCADA       1       SETTLEMENTDATE      DUID  \
9         D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00    BALBG1   
10        D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00    BALBL1   
21        D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00  BBATRYL1   
39        D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00  BOWWBA1G   
40        D  DISPATCH  UNIT_SCADA 1.00000  2024/08/01 00:05:00  BOWWBA1L   
...      ..       ...         ...     ...                  ...       ...   
24261731  D  DISPATCH  UNIT_SCADA 1.00000  2025/02/01 00:00:00   RT_VIC5   
24261732  D  DISPATCH  UNIT_SCADA 1.00000  2025/02/01 00:00:00   RT_VIC6   
24261733  D  DISPATCH  UNIT_SCADA 1.00000  2025/02/01 00:00:00   RT_VIC7   
24261734  D  DISPATCH  UNIT_SCADA 1.00000  2025/02/01 00:00:00   RT_VIC8   
24261735  D  DISPATCH  UNIT_SCADA 1.00000  2025/02/01 00:00:00   RT_VIC9   

          SCADAVALUE          LASTCHANGED  
9            0.00000                  NaN  
10           0.00000                  NaN  
21           0.00000                  NaN  
39           0.00000                  NaN  
40           0.02300                  NaN  
...              ...                  ...  
24261731     0.00000  2025/01/31 23:55:11  
24261732     0.00000  2025/01/31 23:55:11  
24261733     0.00000  2025/01/31 23:55:11  
24261734     0.00000  2025/01/31 23:55:11  
24261735     0.00000  2025/01/31 23:55:11  

[2415745 rows x 8 columns]

<div style="background-color: lightblue; padding: 10px;">

There are 2.4M/24.2M rows of recorded data in UNIT_SCADA data has units that were not in registered participants data.

That maybe labeled "Other" for energy source later.
</div>

#### number of energy sources 

In [86]:
df_participants['Fuel Source - Descriptor'].unique()

array(['Grid', 'Water', 'Solar', 'Diesel', 'Waste Coal Mine Gas', 'Wind',
       'Natural Gas', nan, 'Black Coal', 'Sewerage / Waste Water',
       'Coal Seam Methane', 'Bagasse', 'Landfill Methane / Landfill Gas',
       'Biogas - sludge', 'Solar ', 'Natural Gas / Diesel', 'Ethane',
       'Brown Coal', 'solar', 'Kerosene', '-', 'Natural Gas / Fuel Oil',
       'Bagasse and Diesel'], dtype=object)

In [87]:
df_participants['Fuel Source - Descriptor'].nunique()

22

<div style="background-color: lightblue; padding: 10px;">

There are 22 fuel energy sources.
</div>

### Generator information

#### stats

In [88]:
df_summary.describe()

7  TRANSMISSIONLOSSFACTOR  DISTRIBUTIONLOSSFACTOR  \
count 79627.00000             79627.00000             79627.00000   
mean      7.00000                 0.98193                 0.99926   
std       0.00000                 0.04321                 0.01248   
min       7.00000                 0.67450                 0.82000   
25%       7.00000                 0.96370                 1.00000   
50%       7.00000                 0.98540                 1.00000   
75%       7.00000                 1.00000                 1.00000   
max       7.00000                 1.28410                 1.16755   

       MINIMUM_ENERGY_PRICE  MAXIMUM_ENERGY_PRICE  MIN_RAMP_RATE_UP  \
count           79627.00000           79627.00000       20876.00000   
mean             -981.23834           12274.75436           4.17594   
std                45.67111            3227.75574          10.59900   
min             -1284.10000            3372.50000           0.00000   
25%             -1000.00000            9993.00000           1.00000   
50%              -985.20000           13019.40000           3.00000   
75%              -962.50000           14534.80000           3.00000   
max              -674.50000           19955.25000         288.00000   

       MIN_RAMP_RATE_DOWN  MAX_RAMP_RATE_UP  MAX_RAMP_RATE_DOWN  \
count         20876.00000       36397.00000         36397.00000   
mean              4.17594          75.60656            77.93901   
std              10.59900         164.10121           170.61033   
min               0.00000           0.00000             0.00000   
25%               1.00000           5.00000             5.00000   
50%               3.00000          20.00000            20.00000   
75%               3.00000          57.00000            57.00000   
max             288.00000        1492.00000          1492.00000   

       IS_AGGREGATED  LOAD_MINIMUM_ENERGY_PRICE  LOAD_MAXIMUM_ENERGY_PRICE  \
count    49142.00000                  169.00000                  169.00000   
mean         0.12380                 -970.07988                16976.37698   
std          0.32936                   56.66950                  991.70240   
min          0.00000                -1047.59000                14497.00000   
25%          0.00000                -1000.90000                16674.00000   
50%          0.00000                 -986.60000                17265.50000   
75%          0.00000                 -952.80000                17515.75000   
max          1.00000                 -828.40000                18332.75000   

       LOAD_MIN_RAMP_RATE_UP  LOAD_MIN_RAMP_RATE_DOWN  LOAD_MAX_RAMP_RATE_UP  \
count              169.00000                169.00000              169.00000   
mean                45.29586                 45.29586              121.95266   
std                 56.78894                 56.78894              190.63028   
min                  2.00000                  2.00000                2.00000   
25%                 10.00000                 10.00000               10.00000   
50%                 34.00000                 34.00000               40.00000   
75%                 54.00000                 54.00000              132.00000   
max                288.00000                288.00000              720.00000   

       LOAD_MAX_RAMP_RATE_DOWN  SECONDARY_TLF  
count                169.00000      283.00000  
mean                 121.95266        0.97256  
std                  190.63028        0.04021  
min                    2.00000        0.84230  
25%                   10.00000        0.95715  
50%                   40.00000        0.98750  
75%                  132.00000        1.00040  
max                  720.00000        1.01550

<div style="background-color: lightblue; padding: 10px;">

TRANSMISSIONLOSSFACTOR is energy losses during transmission in the grid. 

DISTRIBUTIONLOSSFACTOR is column represents the factor used to account for energy losses that occur during the distribution of electricity from the generation source to the end consumers.

MINIMUM_ENERGY_PRICE is the minimum price for energy that a unit can offer, adjusted for transmission and distribution loss factors. But they have negative values. It could represent a situation where the energy price is pushed below zero to balance the market.

MIN_RAMP_RATE_UP is the minimum rate at which a unit can increase its output (in MW/min) when transitioning from a low to a high generation state. It indicates how quickly a generator can respond when there is an increased demand for electricity or when it is asked to increase its generation output.

MAX_RAMP_RATE_DOWN is the maximum rate at which a unit can decrease its output (in MW/min).

LOAD_MINIMUM_ENERGY_PRICE also has negative values.

</div>

#### dispatchable types

In [90]:
df_summary.DISPATCHTYPE.unique()

array(['BIDIRECTIONAL', 'GENERATOR', 'LOAD', nan], dtype=object)

<div style="background-color: lightblue; padding: 10px;">

There are many null values.

LOAD: A load unit (consumer).

GENERATOR: A power generation unit.

BIDIRECTIONAL: A unit that can both generate and consume power, such as battery storage or hybrid systems.
</div>

#### regions

In [91]:
df_summary.REGIONID.unique()

array(['SA1', 'NSW1', 'VIC1', 'QLD1', 'TAS1', 'SNOWY1', nan], dtype=object)

<div style="background-color: lightblue; padding: 10px;">

There is SNOWY1 value, The Snowy Mountains Hydroelectric Scheme (SMHS) is a large-scale hydroelectric power system in southeastern Australia, primarily located in the Snowy Mountains. It spans for ACT, VIC, NSW.
</div>

#### number of stations

In [93]:
df_summary.STATIONID.nunique()

585

<div style="background-color: lightblue; padding: 10px;">

There are 585 stations.
</div>

#### check number of dispatchable units

In [94]:
df_summary.DUID.nunique()

933

<div style="background-color: lightblue; padding: 10px;">

There are over 900 dispatchable units in 585 stations. But there are only 500 units appearing in UNIT_SCADA.
</div>

In [98]:
summary_ids = df_summary.DUID.unique()
summary_ids_scada = df_unit_scada.DUID.unique()

In [99]:
diff_lst = list(set(summary_ids_scada) - set(summary_ids))
print(diff_lst)  

[]


<div style="background-color: lightblue; padding: 10px;">

There is no DUID in UNIT_SCADA that are not in SUMMARY file.
</div>

### Demand data

#### stats

In [100]:
df_demand.describe()

3  OPERATIONAL_DEMAND.1  OPERATIONAL_DEMAND_ADJUSTMENT  \
count 44160.00000           44160.00000                    44160.00000   
mean      3.00000            4046.07158                        0.00000   
std       0.00000            2671.62785                        0.00000   
min       3.00000            -205.00000                        0.00000   
25%       3.00000            1259.00000                        0.00000   
50%       3.00000            4403.00000                        0.00000   
75%       3.00000            6245.00000                        0.00000   
max       3.00000           12463.00000                        0.00000   

       WDR_ESTIMATE  
count   44160.00000  
mean        0.00876  
std         0.28772  
min         0.00000  
25%         0.00000  
50%         0.00000  
75%         0.00000  
max        14.00000

In [102]:
df_demand[df_demand['OPERATIONAL_DEMAND.1'] < 0].shape[0]

36

<div style="background-color: lightblue; padding: 10px;">

There are 36 rows of OPERATIONAL_DEMAND have negative values. They may be errors. 

All OPERATIONAL_DEMAND_ADJUSTMENT values are all 0.
</div>

## 2D Exploration

Explore the file [EDA]2DDataExploration_and_DeepAnalysis because the content may be too long.